### ISSUE TITLE
combine jsmn and jscon_scanf to implement a more robust json de-serialization function

### ISSUE THREAD

**Issue by mlite** [cee-studio/libdiscord#1]
jsmn can find all the JSON tokens with minimum memory usage. 
jscon_scanf use C's scanf format string  convention to de-serialize json string as C struct instance, however it does not support de-serialize nested json objects yet. 

we can utilize the strength of both jsmn an jscon_scanf to implement a new function that can de-serialize arbitrary nested json objects to the corresponding C structs.

**Comment by LucasMull** [comment]
Ok. I'll begin developing a combination of jsmn and jscon_scanf once I get a simple API transfer working.

**Comment by LucasMull** [comment]
But I have a question, what do you mean by support nested objects? I have updated jscon_scanf to allow for nested object members de-serialization.

```c
jscon_item_t *item;
char *string;
bool boolean;
int number;

char buffer[] = "{\"alpha\":[1,2,3,4], \"beta\":\"This is a string.", \"gamma\":true, \"omega\":{\"number\":1}}";
/* order of arguments doesn't have to be the same as the json string */
jscon_scanf(buffer, "%s[beta] %b[gamma] %ji[alpha] %d[omega][number]", string, &boolean, &item, &number);
```

%d[omega][number] 
Isn't this what we want? I tested with deeper nesting and works fine too

**Comment by mlite** [comment]
> But I have a question, what do you mean by support nested objects? I have updated jscon_scanf to allow for nested object members de-serialization.
> 
> ```c
> jscon_item_t *item;
> char *string;
> bool boolean;
> int number;
> 
> char buffer[] = "{\"alpha\":[1,2,3,4], \"beta\":\"This is a string.", \"gamma\":true, \"omega\":{\"number\":1}}";
> /* order of arguments doesn't have to be the same as the json string */
> jscon_scanf(buffer, "%s[beta] %b[gamma] %ji[alpha] %d[omega][number]", string, &boolean, &item, &number);
> ```
> 
> %d[omega][number]
> Isn't this what we want? I tested with deeper nesting and works fine too

This design provides a great flexibility to extract data based on the need.  There is a use case that needs this design.

**Comment by mlite** [comment]
For a property, if its value is an array of somethings
```
"labels" [ "l1", "l2", "l3", 1, 2, 4 ]
```
we de-serialize  the array as a null terminated C array.  

the element of the above array has a sum type like the following
```
struct {
    enum {
         STRING,
          INT
   }  tag,
   union {
       char * s;
       int  i;
   } _;
}
```

**Comment by LucasMull** [comment]
> For a property, whose value is an array of somethings like the following. 
> 
> ```
> 
> "labels" [ "l1", "l2", "l3", 1, 2, 4 ]
> 
> ```
> 
> we de-serialize  the array as a null terminated C array.  
> 
> 
> 
> the elements of the above array is a pointer of the following sum type
> 
> ```
> 
> struct {
> 
>     enum {
> 
>          STRING,
> 
>          INT
> 
>    }  tag;
> 
>    union {
> 
>        char * s;
> 
>        int  i;
> 
>    } _;
> 
> };
> 
> ```
> 
> 
> 
> This struct definition should be provided by developers. 

Really simple solution, this is the way to go.

**Comment by LucasMull** [comment]
First I will make a single file version of jscon-scanf.c, then work on combining with jsmn. I think it would make sense to dich the 'c' in jscon and simply call it json_scanf instead.

**Comment by mlite** [comment]
> First I will make a single file version of jscon-scanf.c, then work on combining with jsmn. I think it would make sense to ditch the 'c' in jscon and simply call it json_scanf instead.

Yes, it's a very good plan. Taking small steps to gradually involve the code base will make testing much easier.    `json_scanf` is a better name.   Please use our compiler to build it, it will make troubleshooting memory errors much easier.   I will add libcurl and other dependency to the release.
